File: src/test/java/fr/deroffal/aoc/Day04Test.java
package fr.deroffal.aoc;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

public class Day04Test {

	@Test
	void aa_bb_cc_dd_ee_valide() {
		assertTrue(Day04.isLineValid("aa bb cc dd ee"));
	}

	@Test
	void aa_bb_cc_dd_aa_non_valide() {
		assertFalse(Day04.isLineValid("aa bb cc dd aa"));
	}

	@Test
	void aa_bb_cc_dd_aaa_valide() {
		assertTrue(Day04.isLineValid("aa bb cc dd aaa"));
	}

	@Test
	void abcde_fghij_valide_anagram() {
		assertTrue(Day04.isLineValidWithAnagram("abcde fghij"));
	}

	@Test
	void abcde_xyz_ecdab_non_valide_anagram() {
		assertFalse(Day04.isLineValidWithAnagram("abcde xyz ecdab"));
	}

	@Test
	void a_ab_abc_abd_abf_abj_valide_anagram() {
		assertTrue(Day04.isLineValidWithAnagram("a ab abc abd abf abj"));
	}

	@Test
	void iiii_oiii_ooii_oooi_oooo_valide_anagram() {
		assertTrue(Day04.isLineValidWithAnagram("iiii oiii ooii oooi oooo"));
	}

	@Test
	void oiii_ioii_iioi_iiio_non_valide_anagram() {
		assertFalse(Day04.isLineValidWithAnagram("oiii ioii iioi iiio"));
	}
}

File: kotlin/src/test/kotlin/fr/deroffal/aoc/kotlin/Day04Test.kt
package fr.deroffal.aoc.kotlin

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

internal class Day04Test {

    private val day04 = Day04()

    @Test
    fun `part1 matches examples`() {
        val input: List<String> = listOf("aa bb cc dd ee", "aa bb cc dd aa", "aa bb cc dd aaa")
        assertEquals(2, day04.solvePart1(input))
    }

    @Test
    fun `part2 matches examples`() {
        val input: List<String> = listOf("abcde fghij", "abcde xyz ecdab", "a ab abc abd abf abj", "iiii oiii ooii oooi oooo", "oiii ioii iioi iiio")
        assertEquals(3, day04.solvePart2(input))
    }
}